src/typemap.c

```diff
@@ -47,8 +47,9 @@ static int sig_match_by_type_simple(jl_value_t **types, size_t n, jl_tupletype_t
     for (i = 0; i < lensig; i++) {
         jl_value_t *decl = jl_field_type(sig, i);
         jl_value_t *a = types[i];
-        if (jl_is_type_type(decl)) {
-            jl_value_t *tp0 = jl_tparam0(decl);
+        jl_value_t *unw = jl_is_unionall(decl) ? ((jl_unionall_t*)decl)->body : decl;
```

Comment:
the support here should agree with the computation in `jl_typemap_insert` and be duplicated in `sig_match_simple` – it doesn't matter which way it goes, but they just need to be kept in sync